hastexo/hastexo.py

```diff
@@ -253,47 +258,33 @@ def get_configuration(self):
         settings = self.get_xblock_settings(default=DEFAULT_SETTINGS)
         return get_xblock_configuration(settings, self.provider)
 
-    def stack_set(self, prop, value):
-        if not self.stacks.get(self.stack_name):
-            self.stacks[self.stack_name] = {}
-
-        self.stacks[self.stack_name][prop] = value
-
-    def stack_get(self, l1=None, l2=None):
-        retval = self.stacks.get(self.stack_name)
-        if retval and l1:
-            retval = retval.get(l1)
-            if retval and l2:
-                retval = retval.get(l2)
+    @transaction.atomic()
+    def stack_update(self, data):
+        course_id, student_id = self.get_block_ids()
+        stack, _ = Stack.objects.select_for_update().get_or_create(
+            student_id = student_id,
+            course_id = course_id,
+            name = self.stack_name
+        )
+        stack.__dict__.update(data)
+        stack.save()
+
+    @transaction.atomic()
+    def stack_get(self, prop = None):
+        course_id, student_id = self.get_block_ids()
+        stack, _ = Stack.objects.select_for_update().get_or_create(
+            student_id = student_id,
+            course_id = course_id,
+            name = self.stack_name
+        )
 
-        return retval
+        if prop:
+            return getattr(stack, prop)
+        else:
+            return stack
 
-    def suspend_user_stack(self, configuration):
-        suspend_timeout = configuration.get("suspend_timeout")
-        if suspend_timeout:
-            # If the suspend task is pending, revoke it.
-            stack_suspend_id = self.stack_get("suspend_id")
-            if stack_suspend_id:
-                logger.info(
-                    'Revoking suspend task [%s] for [%s]' % (stack_suspend_id,
-                                                             self.stack_name)
-                )
-                from lms import CELERY_APP
-                CELERY_APP.control.revoke(stack_suspend_id)
-                self.stack_set("suspend_id", None)
-
-            # (Re)schedule the suspension in the future.
-            args = (configuration, self.stack_name)
-            task = SuspendStackTask()
-            result = task.apply_async(args=args, countdown=suspend_timeout)
-            logger.info(
-                'Scheduled suspend task [%s] '
-                'for [%s] in %s seconds' % (result.id,
-                                            self.stack_name,
-                                            suspend_timeout)
-            )
-            self.stack_set("suspend_id", result.id)
-            self.stack_set("suspend_timestamp", int(time.time()))
+    def reset_suspend_timestamp(self):
+        self.stack_update({"suspend_timestamp": timezone.now()})
```

Comment:
This is a very good point, thanks for bringing it up.  I did some research and apparently (who'd'a thunk!), Django does The Right Thing.  Datetime objects are always stored in UTC in the database, and:

"If USE_TZ is True, this will be an aware datetime representing the current time in UTC. Note that now() will always return times in UTC regardless of the value of TIME_ZONE; you can use localtime() to get the time in the current time zone."  (https://docs.djangoproject.com/en/2.0/ref/utils/#django.utils.timezone.now)

`USE_TZ` is set to True, btw.